### PR TITLE
Add Jetpack navigation module

### PR DESCRIPTION
**Pull request by marukami**
Adding support for the Navigation Graph #305 is a little tricky. The issue you run into is cause we need the navigation controller, and, so this means when restoring from process death, the fragments are restored in the activity#onCreate super before the activity has inflated its content. So, the `NavigationController` is not always accessible in Fragment#onCreate. When not restoring from process death, the existing ViewModel delegates method of creating the ViewModel like `fragmentViewModel` works fine.

For now, my current workaround is to remove the lifecycle aware lazy and use Lazy. I think it should be possible to make navigation ViewModel delegate auto-subscribe by delaying the subscription until `onStart` when the `NavigationController` is not yet accessible.

I still need to add some CI tests, but, I thought I would open the PR now to get some feedback on my current approach and if it makes sense to have navigation as an add-on module.

### CODE REVIEW

**Comment by nealsanche** [comment]
Personally, I like this direction. We were definitely having difficulties with restoring after process death, and if this solves that issue, it'd be very welcome.

File: mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelFactory.kt [hunk 1]
```diff
@@ -101,11 +101,13 @@ data class FragmentViewModelContext(
     /**
      * The fragment owner of the ViewModel.
      */
-    val fragment: Fragment
+    val fragment: Fragment,
+    private val ownerProvider: ViewModelStoreOwner = fragment,
+    private val savedStateRegistryProvider: SavedStateRegistry = fragment.savedStateRegistry
```
Comment (mradzinski) [inline mvrx/src/main/kotlin/com/airbnb/mvrx/MvRxViewModelFactory.kt]: Shouldn't this be `private val savedStateRegistryProvider: SavedStateRegistry = ownerProvider.savedStateRegistry`? Just wondering what would happen If you assign an `ownerProvider` but not a `savedStateRegistryProvider` .

**Comment by kdreamix** [comment]
Illegal state exception will still be thrown in this approach on configuration changed

**Comment by marukami** [comment]
> Illegal state exception will still be thrown in this approach on configuration changed

@kdreamix Good to know thanks for the post.

**Comment by kdreamix** [comment]
Any idea on how to work around that?

**Comment by marukami** [comment]
> Any idea on how to work around that?

At the time you asked no. I think I have a working prototype now. 

Sorry I Accidentally deleted this PR 😅 

The MvRx view needs to be subscribed to the ViewModel and there seems to be two ways to trigger this safely. 
1. `try/catch` in the `lifecyleOwner` onCreate, and fallback to onStart if `onCreate` fails
2. Access the ViewModel from the Fragment in or after and only after the `onActivityCreated`, forcing the lazy initializer to fire.

I opted to use both as some fragments might only ever use invalidate.

**Pull request closed without merging**